src/core/ext/xds/xds_client.cc

```diff
@@ -1565,6 +1600,56 @@ RefCountedPtr<XdsClient::XdsChannel> XdsClient::GetOrCreateXdsChannelLocked(
   return xds_channel;
 }
 
+bool XdsClient::HasUncachedResources(const AuthorityState& authority_state) {
+  for (const auto& type_resource : authority_state.resource_map) {
+    for (const auto& key_state : type_resource.second) {
+      if (key_state.second.meta.client_status ==
+          XdsApi::ResourceMetadata::REQUESTED) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool XdsClient::MaybeFallbackLocked(const std::string& authority,
+                                    AuthorityState& authority_state) {
+  if (!HasUncachedResources(authority_state)) {
+    return false;
+  }
+  std::vector<const XdsBootstrap::XdsServer*> xds_servers;
+  if (authority != kOldStyleAuthority) {
+    xds_servers = bootstrap().LookupAuthority(authority)->servers();
+  }
+  if (xds_servers.empty()) xds_servers = bootstrap().servers();
+  for (size_t i = authority_state.xds_channels.size(); i < xds_servers.size();
+       ++i) {
+    authority_state.xds_channels.emplace_back(
+        GetOrCreateXdsChannelLocked(*xds_servers[i], "fallback"));
+    for (const auto& type_resource : authority_state.resource_map) {
+      for (const auto& key_state : type_resource.second) {
+        gpr_log(GPR_INFO, "[xds_client %p] xds server %s, subscribing to %s %s",
+                this, xds_servers[i]->server_uri().c_str(),
+                std::string(type_resource.first->type_url()).c_str(),
+                key_state.first.id.c_str());
+        authority_state.xds_channels.back()->SubscribeLocked(
+            type_resource.first, {authority, key_state.first});
+      }
+    }
+    if (authority_state.xds_channels.back()->status().ok()) {
+      gpr_log(GPR_INFO, "[xds_client %p] Performed fallback to %s", this,
+              xds_servers[i]->server_uri().c_str());
+      return true;
+    } else {
+      gpr_log(GPR_ERROR, "Fallback channel %s is in state %s",
```

Comment:
Instead of having two different messages here, I suggest something like this:

```
if (GRPC_TRACE_FLAG_ENABLED(grpc_xds_client_trace)) {
  gpr_log(GPR_INFO, "[xds_client %p] authority %s: added fallback server %s (%s)",
          this, authority.c_str(), xds_servers[i]->server_uri().c_str(),
          authority_state.xds_channels.back()->status().ToString().c_str());
}
if (authority_state.xds_channels.back()->status().ok()) return tre;
```